sdk/go/pulumi-language-go/main.go

```diff
@@ -218,41 +219,59 @@ func newLanguageHost(engineAddress, cwd, tracing, binary, buildTarget string) pu
 }
 
 // modInfo is the useful portion of the output from `go list -m -json all`
-// with respect to plugin acquisition
+// with respect to plugin acquisition.
 type modInfo struct {
-	Path    string
+	// Path is the module import path.
+	Path string
+
+	// Version of the module.
 	Version string
-	Dir     string
+
+	// Dir is the directory holding the source code of the module.
+	//
+	// This is empty in vendor mode.
+	Dir string
 }
 
-// Returns the pulumi-plugin.json if found. If not found, then returns nil, nil.
-// The lookup path for pulumi-plugin.json is
-// 1. m.Dir
-// 2. m.Dir/go
-// 3. m.Dir/go/*
-func (m *modInfo) readPulumiPluginJSON() (*plugin.PulumiPluginJSON, error) {
-	if m.Dir == "" {
-		return nil, nil
+// Returns the pulumi-plugin.json if found.
+// If not found, then returns nil, nil.
+//
+// The lookup path for pulumi-plugin.json is:
+//
+//   - m.Dir
+//   - m.Dir/go
+//   - m.Dir/go/*
+//
+// moduleRoot is the root directory of the module that imports this plugin.
+// It is used to resolve the vendor directory.
+// moduleRoot may be empty if unknown.
+func (m *modInfo) readPulumiPluginJSON(moduleRoot string) (*plugin.PulumiPluginJSON, error) {
+	dir := m.Dir
+	if dir == "" && moduleRoot != "" {
+		// m.Dir is empty in vendor mode.
+		//
+		// In this case, we use the vendor directory to find the plugin
+		// if we know the root of the module that imports this plugin.
+		dir = filepath.Join(moduleRoot, "vendor", filepath.ToSlash(m.Path))
```

Comment:
moduleRoot will never be empty—it's the Pulumi program root.
I allowed for it to be empty to avoid messing with some of the tests; I'll drop that and add a contract assert instead.